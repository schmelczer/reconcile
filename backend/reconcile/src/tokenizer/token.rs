#[derive(Debug, Clone)]
pub struct Token<T>
where
    T: PartialEq + Clone,
{
    normalised: T,
    original: String,
}

impl From<&str> for Token<String> {
    fn from(s: &str) -> Self {
        Token {
            normalised: s.to_string(),
            original: s.to_string(),
        }
    }
}

impl<T> Token<T>
where
    T: PartialEq + Clone,
{
    pub fn new(normalised: T, original: String) -> Self {
        Token {
            normalised,
            original,
        }
    }

    pub fn original(&self) -> &str {
        &self.original
    }

    pub fn normalised(&self) -> &T {
        &self.normalised
    }

    pub fn get_original_length(&self) -> usize {
        self.original.chars().count()
    }
}

impl<T> PartialEq for Token<T>
where
    T: PartialEq + Clone,
{
    fn eq(&self, other: &Self) -> bool {
        self.normalised == other.normalised
    }
}
