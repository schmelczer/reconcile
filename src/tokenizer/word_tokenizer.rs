use super::token::Token;

/// Splits text on word boundaries, creating tokens of alternating words and
/// whitespace with the whitespace getting unique IDs.
///
/// ## Example
///
/// ```not_rust
/// "Hi there!" -> ["Hi", " ", "there!"]
/// ```
pub fn word_tokenizer(text: &str) -> Vec<Token<String>> {
    let mut result = Vec::new();

    let mut previous_boundary_index = 0;
    let mut previous_char_is_whitespace = text.chars().next().is_none_or(char::is_whitespace);

    for (i, c) in text.char_indices() {
        let is_current_char_whitespace = c.is_whitespace();
        if previous_char_is_whitespace != is_current_char_whitespace {
            result.push(text[previous_boundary_index..i].into());
            previous_boundary_index = i;
        }

        previous_char_is_whitespace = is_current_char_whitespace;
    }

    if previous_boundary_index < text.len() {
        result.push(text[previous_boundary_index..].into());
    }

    if result.is_empty() {
        return result;
    }

    // normalize whitespace tokens by concatenating with the following token
    for i in 0..result.len() - 1 {
        if result[i].original().chars().all(char::is_whitespace) {
            let normalized = result[i].normalized().to_owned() + result[i + 1].original();
            result[i].set_normalized(normalized);
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn test_with_snapshots() {
        assert_debug_snapshot!(word_tokenizer("Hi there!"));

        assert_debug_snapshot!(word_tokenizer(""));

        assert_debug_snapshot!(word_tokenizer(" what? "));

        assert_debug_snapshot!(word_tokenizer(" hello, \nwhere are you?"));
    }
}
