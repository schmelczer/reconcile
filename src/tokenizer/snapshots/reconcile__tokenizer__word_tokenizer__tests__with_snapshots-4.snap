---
source: src/tokenizer/word_tokenizer.rs
expression: "word_tokenizer(\" hello, \\nwhere are you?\")"
snapshot_kind: text
---
[
    Token {
        normalized: " hello,",
        original: " ",
        is_left_joinable: true,
        is_right_joinable: true,
    },
    Token {
        normalized: "hello,",
        original: "hello,",
        is_left_joinable: true,
        is_right_joinable: true,
    },
    Token {
        normalized: " \nwhere",
        original: " \n",
        is_left_joinable: true,
        is_right_joinable: true,
    },
    Token {
        normalized: "where",
        original: "where",
        is_left_joinable: true,
        is_right_joinable: true,
    },
    Token {
        normalized: " are",
        original: " ",
        is_left_joinable: true,
        is_right_joinable: true,
    },
    Token {
        normalized: "are",
        original: "are",
        is_left_joinable: true,
        is_right_joinable: true,
    },
    Token {
        normalized: " you?",
        original: " ",
        is_left_joinable: true,
        is_right_joinable: true,
    },
    Token {
        normalized: "you?",
        original: "you?",
        is_left_joinable: true,
        is_right_joinable: true,
    },
]
